pkg/networking/middleware/response.go

```diff
@@ -0,0 +1,123 @@
+package middleware
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/snyk/error-catalog-golang-public/snyk"
+	"github.com/snyk/error-catalog-golang-public/snyk_errors"
+)
+
+type ResponseMiddleware struct {
+	next       http.RoundTripper
+	errHandler func(error, context.Context) error
+}
+
+func NewReponseMiddleware(roundTriper http.RoundTripper, errHandler func(error, context.Context) error) *ResponseMiddleware {
+	return &ResponseMiddleware{
+		next:       roundTriper,
+		errHandler: errHandler,
+	}
+}
+
+func errFromResponse(res *http.Response) error {
+	if res.StatusCode >= 200 && res.StatusCode < 300 {
+		return nil
+	}
+
+	if err := errFromStatusCode(res.StatusCode); err != nil {
+		return err
+	}
+
+	if err := errFromJSONAPI(res); err != nil {
+		return err
+	}
+
+	return snyk_errors.Error{
```

Comment:
Suggestion: catching unhandled cases as an error seems like a very strong approach. I think we might want to be less strong, as we don't want to change the default behaviour too much.